### PR TITLE
Hash table more tasks

### PR DESCRIPTION
**Pull request by danblitzhou**
- hash_table_clear
- hash_table_init-unbounded and add memcpy unroll bound for hash_table_init-bounded so that array memory model works
- _possible real sat cases_: according to description of `hash_iter_begin`, if there are no items in the table (`p_impl->entry_count == 0`), "the returned iterator will return true for aws_hash_iter_done(&iter)";
- the above issue in `hash_iter_begin` leads to similar issue for `hash_table_foreach`, which uses `aws_hash_iter_*` functions to iterate all items in the hash_table and apply a callback on them; if the callback returns a value that triggers deletion of item, it is possible then that: entry_count == 0  --> hash_table_foreach iterates anyways and calls `aws_hash_iter_delete`, which operates `entry_count -= 1` ---> entry_count == -1

### CODE REVIEW

File: seahorn/jobs/hash_table_foreach/aws_hash_table_foreach_harness.c [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+int main(void) {
+  struct aws_hash_table map;
+  initialize_bounded_aws_hash_table(&map, MAX_TABLE_SIZE);
+  ensure_hash_table_has_valid_destroy_functions(&map);
+  map.p_impl->equals_fn = nondet_equals;
+  map.p_impl->hash_fn = uninterpreted_hasher;
+  assume(aws_hash_table_is_valid(&map));
+
+  void *context;
```
Comment (agurfinkel) [inline seahorn/jobs/hash_table_foreach/aws_hash_table_foreach_harness.c:23]: better initialize this pointer. Nothing happens to it, but might as well initialize to something like `nd_voidp()`

File: seahorn/jobs/hash_table_init_unbounded/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+add_executable(hash_table_init_unbounded
+  aws_hash_table_init_unbounded_harness.c)
+set(MAX_TABLE_SIZE 8)
+math(EXPR MEMCPY_UNROLL_BND "${MAX_TABLE_SIZE} * 64")
```
Comment (agurfinkel) [inline seahorn/jobs/hash_table_init_unbounded/CMakeLists.txt:4]: maybe this is a copy-paste error? the name of the task is `unbounded`, so why do we bound?

File: seahorn/lib/utils.c [hunk 3]
```diff
@@ -101,4 +101,12 @@ void assert_byte_cursor_equivalence(
...
 }
 
-bool uninterpreted_predicate_fn(uint8_t value);
\ No newline at end of file
+bool uninterpreted_predicate_fn(uint8_t value);
+
+bool nondet_equals(const void *const a, const void *const b) {
+  sassert(a != NULL);
+  sassert(b != NULL);
+  return nd_bool();
```
Comment (agurfinkel) [inline seahorn/lib/utils.c:109]: ```return a == b ? true : nd_bool()```

does not make any difference but makes me feel better :)

**Comment by danblitzhou** [comment]
The pre-condition `aws_hash_table_entry_count_is_valid`: `hash_table.p_impl->entry_count` matches actual number of non-zero element hash_codes is necessary for `hash_table.p_impl->entry_count` to be updated correctly after a `foreach` iteration that could potentially delete elements; but it's also very expensive to add this pre-condition (loop, high symmetry), and does not really affect memory safety;
Therefore we shall use two versions of `hash_table_foreach`:
a. strictly enforce `aws_hash_table_entry_count_is_valid` at a cost of verification time
b. stub `hash_table_state_is_valid` further and ignore the post-condition on entry_count to achieve reasonable verification time and check memory safety only. @agurfinkel

File: seahorn/jobs/hash_table_foreach/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+add_executable(hash_table_foreach
+  ${AWS_C_COMMON_ROOT}/source/common.c
+  aws_hash_table_foreach_harness.c)
+set(MAX_TABLE_SIZE 2)
```
Comment (agurfinkel) [inline seahorn/jobs/hash_table_foreach/CMakeLists.txt:4]: I think the `MAX_TABLE_SIZE` bound and `UNROLL_BOUND` can be relaxed for the non-deep version of the verification task.

File: seahorn/jobs/hash_table_foreach_deep/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+add_executable(hash_table_foreach_deep
+  ${AWS_C_COMMON_ROOT}/source/common.c
+  aws_hash_table_foreach_deep_harness.c)
+set(MAX_TABLE_SIZE 2)
```
Comment (agurfinkel) [inline seahorn/jobs/hash_table_foreach_deep/CMakeLists.txt:4]: what is the verification time. Can we go to 3, 4, ?

**Review by agurfinkel — APPROVED** [review]

**Pull request merged**